File: android/app/src/main/java/com/smartenergy/tracker/network/SocketManager.java
package com.smartenergy.tracker.network;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.google.gson.Gson;
import com.smartenergy.tracker.model.AlertItem;
import com.smartenergy.tracker.model.Appliance;
import com.smartenergy.tracker.model.Telemetry;

import org.json.JSONException;
import org.json.JSONObject;

import java.net.URISyntaxException;

import io.socket.client.IO;
import io.socket.client.Socket;

public class SocketManager {
    private static final String TAG = "SocketManager";
    private static SocketManager instance;
    private Socket socket;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final Gson gson = new Gson();
    private boolean isConnected = false;

    public interface SocketEventListener {
        void onConnected();
        void onDisconnected();
        void onTelemetryUpdate(Telemetry telemetry);
        void onApplianceStateChanged(Appliance appliance);
        void onAlertReceived(AlertItem alert);
    }

    private SocketEventListener listener;

    private SocketManager() {}

    public static synchronized SocketManager getInstance() {
        if (instance == null) {
            instance = new SocketManager();
        }
        return instance;
    }

    public void setListener(SocketEventListener listener) {
        this.listener = listener;
    }

    public void connect(Context context) {
        disconnect();

        PreferencesManager prefs = new PreferencesManager(context);
        String serverUrl = prefs.getBaseUrl();

        try {
            IO.Options options = new IO.Options();
            options.reconnection = true;
            options.reconnectionAttempts = 30;
            options.reconnectionDelay = 1000;
            options.timeout = 10000;

            socket = IO.socket(serverUrl, options);

            socket.on(Socket.EVENT_CONNECT, args -> {
                Log.d(TAG, "Socket.IO Connected to " + serverUrl);
                isConnected = true;
                if (listener != null) {
                    mainHandler.post(() -> listener.onConnected());
                }
            });

            socket.on(Socket.EVENT_DISCONNECT, args -> {
                Log.d(TAG, "Socket.IO Disconnected");
                isConnected = false;
                if (listener != null) {
                    mainHandler.post(() -> listener.onDisconnected());
                }
            });

            socket.on(Socket.EVENT_CONNECT_ERROR, args -> {
                Log.e(TAG, "Socket.IO Connect Error: " + (args.length > 0 ? args[0] : "unknown"));
                isConnected = false;
                if (listener != null) {
                    mainHandler.post(() -> listener.onDisconnected());
                }
            });

            // Initial snapshot on connect
            socket.on("init:snapshot", args -> {
                if (args.length > 0 && listener != null) {
                    try {
                        String jsonString = args[0].toString();
                        Telemetry telemetry = gson.fromJson(jsonString, Telemetry.class);
                        mainHandler.post(() -> listener.onTelemetryUpdate(telemetry));
                    } catch (Exception e) {
                        Log.e(TAG, "Snapshot parse error: " + e.getMessage());
                    }
                }
            });

            // Continuous telemetry stream (1 Hz)
            socket.on("telemetry:update", args -> {
                if (args.length > 0 && listener != null) {
                    try {
                        String jsonString = args[0].toString();
                        Telemetry telemetry = gson.fromJson(jsonString, Telemetry.class);
                        mainHandler.post(() -> listener.onTelemetryUpdate(telemetry));
                    } catch (Exception e) {
                        Log.e(TAG, "Telemetry parse error: " + e.getMessage());
                    }
                }
            });

            socket.on("appliance:state_changed", args -> {
                if (args.length > 0 && listener != null) {
                    try {
                        String jsonString = args[0].toString();
                        Appliance appliance = gson.fromJson(jsonString, Appliance.class);
                        mainHandler.post(() -> listener.onApplianceStateChanged(appliance));
                    } catch (Exception e) {
                        Log.e(TAG, "Appliance change parse error: " + e.getMessage());
                    }
                }
            });

            socket.on("alert:new", args -> {
                if (args.length > 0 && listener != null) {
                    try {
                        String jsonString = args[0].toString();
                        AlertItem alert = gson.fromJson(jsonString, AlertItem.class);
                        mainHandler.post(() -> listener.onAlertReceived(alert));
                    } catch (Exception e) {
                        Log.e(TAG, "Alert parse error: " + e.getMessage());
                    }
                }
            });

            socket.connect();

        } catch (URISyntaxException e) {
            Log.e(TAG, "Socket connection error: " + e.getMessage());
        }
    }

    public void disconnect() {
        if (socket != null) {
            socket.disconnect();
            socket.off();
            socket = null;
            isConnected = false;
        }
    }

    public boolean isConnected() {
        return isConnected;
    }

    public void toggleAppliance(String id, boolean targetState) {
        if (socket != null && isConnected) {
            try {
                JSONObject obj = new JSONObject();
                obj.put("id", id);
                obj.put("state", targetState);
                socket.emit("appliance:toggle", obj);
            } catch (JSONException e) {
                Log.e(TAG, "Error emitting toggle: " + e.getMessage());
            }
        }
    }

    public void injectAnomaly(String id, boolean isAnomaly) {
        if (socket != null && isConnected) {
            try {
                JSONObject obj = new JSONObject();
                obj.put("id", id);
                obj.put("isAnomaly", isAnomaly);
                socket.emit("simulation:inject_anomaly", obj);
            } catch (JSONException e) {
                Log.e(TAG, "Error emitting anomaly: " + e.getMessage());
            }
        }
    }
}

File: android/app/src/main/java/com/smartenergy/tracker/network/SocketManager.kt
package com.smartenergy.tracker.network

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.google.gson.Gson
import com.smartenergy.tracker.model.AlertItem
import com.smartenergy.tracker.model.Appliance
import com.smartenergy.tracker.model.Telemetry
import io.socket.client.IO
import io.socket.client.Socket
import org.json.JSONObject
import java.net.URI

class SocketManager private constructor(private val context: Context) {
    private var socket: Socket? = null
    private val gson = Gson()
    private val mainHandler = Handler(Looper.getMainLooper())

    var isConnected: Boolean = false
        private set

    // Callbacks
    var onConnectionChanged: ((Boolean) -> Unit)? = null
    var onTelemetryReceived: ((Telemetry) -> Unit)? = null
    var onApplianceChanged: ((Appliance) -> Unit)? = null
    var onAlertReceived: ((AlertItem) -> Unit)? = null

    companion object {
        private const val TAG = "SocketManager"

        @Volatile
        private var INSTANCE: SocketManager? = null

        fun getInstance(context: Context): SocketManager {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: SocketManager(context.applicationContext).also { INSTANCE = it }
            }
        }
    }

    fun connect() {
        val prefs = PreferencesManager.getInstance(context)
        val url = prefs.socketUrl
        Log.d(TAG, "Connecting to Socket.IO at $url")

        try {
            socket?.disconnect()
            socket?.off()

            val opts = IO.Options().apply {
                reconnection = true
                reconnectionAttempts = 50
                reconnectionDelay = 2000
                timeout = 10000
                forceNew = true
            }

            socket = IO.socket(URI.create(url), opts)

            socket?.on(Socket.EVENT_CONNECT) {
                Log.d(TAG, "Socket connected: ${socket?.id()}")
                isConnected = true
                mainHandler.post { onConnectionChanged?.invoke(true) }
            }

            socket?.on(Socket.EVENT_DISCONNECT) {
                Log.d(TAG, "Socket disconnected")
                isConnected = false
                mainHandler.post { onConnectionChanged?.invoke(false) }
            }

            socket?.on(Socket.EVENT_CONNECT_ERROR) { args ->
                Log.e(TAG, "Socket connect error: ${args.firstOrNull()}")
                isConnected = false
                mainHandler.post { onConnectionChanged?.invoke(false) }
            }

            // Real-time telemetry tick stream (every 1s)
            socket?.on("telemetry:update") { args ->
                if (args.isNotEmpty() && args[0] != null) {
                    try {
                        val json = args[0].toString()
                        val telemetry = gson.fromJson(json, Telemetry::class.java)
                        mainHandler.post { onTelemetryReceived?.invoke(telemetry) }
                    } catch (e: Exception) {
                        Log.e(TAG, "Error parsing telemetry:update", e)
                    }
                }
            }

            // Snapshot upon connection
            socket?.on("init:snapshot") { args ->
                if (args.isNotEmpty() && args[0] != null) {
                    try {
                        val json = args[0].toString()
                        val telemetry = gson.fromJson(json, Telemetry::class.java)
                        mainHandler.post { onTelemetryReceived?.invoke(telemetry) }
                    } catch (e: Exception) {
                        Log.e(TAG, "Error parsing init:snapshot", e)
                    }
                }
            }

            // Appliance relay status change broadcast
            socket?.on("appliance:state_changed") { args ->
                if (args.isNotEmpty() && args[0] != null) {
                    try {
                        val json = args[0].toString()
                        val appliance = gson.fromJson(json, Appliance::class.java)
                        mainHandler.post { onApplianceChanged?.invoke(appliance) }
                    } catch (e: Exception) {
                        Log.e(TAG, "Error parsing appliance:state_changed", e)
                    }
                }
            }

            // Anomaly push alert
            socket?.on("alert:new") { args ->
                if (args.isNotEmpty() && args[0] != null) {
                    try {
                        val json = args[0].toString()
                        val alert = gson.fromJson(json, AlertItem::class.java)
                        mainHandler.post { onAlertReceived?.invoke(alert) }
                    } catch (e: Exception) {
                        Log.e(TAG, "Error parsing alert:new", e)
                    }
                }
            }

            socket?.connect()

        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize socket", e)
            isConnected = false
            mainHandler.post { onConnectionChanged?.invoke(false) }
        }
    }

    fun toggleAppliance(id: String, state: Boolean) {
        try {
            val payload = JSONObject().apply {
                put("id", id)
                put("state", state)
            }
            socket?.emit("appliance:toggle", payload)
            Log.d(TAG, "Emitted appliance:toggle for $id -> $state")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to emit appliance:toggle", e)
        }
    }

    fun disconnect() {
        socket?.disconnect()
        socket?.off()
        isConnected = false
    }
}
